### PR TITLE
feat: whitelist proposal validation

### LINKED ISSUE #473: Whitelist proposal validation strategy
We need to add a whitelist proposal validation strategy

### PR DESCRIPTION
**Pull request by pscott**
Based on the merkle implementation of the voting strategy.

Closes #473

### CODE REVIEW

**Comment by Orland0x** [comment]
hard to see the actual diff cuz of changing paths. I just need to review `starknet/src/proposal_validation_strategies/merkle_whitelist.cairo` and `starknet/src/tests/proposal_validation_strategies/test_merkle_whitelist.cairo` the rest didnt change right?

File: starknet/src/tests/proposal_validation_strategies/test_merkle_whitelist.cairo [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+        let proof = generate_proof(merkle_data.span(), index);
+
+        let mut params = ArrayTrait::<felt252>::new();
+        root.serialize(ref params);
+        threshold.serialize(ref params);
+
+        let mut user_params = ArrayTrait::<felt252>::new();
+        let fake_leaf = Leaf {
+            address: UserAddress::Starknet(contract_address_const::<0x1337>()),
+            voting_power: leaf.voting_power + 1,
```
Comment (Orland0x) [inline starknet/src/tests/proposal_validation_strategies/test_merkle_whitelist.cairo:193]: looks like the voting power and address are both modified

File: starknet/src/proposal_validation_strategies/merkle_whitelist.cairo [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    use option::OptionTrait;
+    use sx::utils::merkle::{assert_valid_proof, Leaf};
+
+    const LEAF_SIZE: usize = 4; // Serde::<Leaf>::serialize().len()
+
+    #[storage]
+    struct Storage {}
+
+    #[derive(Drop, Serde)]
+    struct Input {
```
Comment (Orland0x) [inline starknet/src/proposal_validation_strategies/merkle_whitelist.cairo:16]: could call something like `StrategyParams`, which we could use in all strategies to define the deserialization of `params` so its clear what this refers to.

File: starknet/src/tests/proposal_validation_strategies/test_merkle_whitelist.cairo [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+    use starknet::syscalls::deploy_syscall;
+    use starknet::SyscallResult;
+    use result::ResultTrait;
+    use option::OptionTrait;
+    use traits::TryInto;
+    use serde::Serde;
+    use starknet::contract_address_const;
+    use sx::types::UserAddress;
+
+    // Checks that that the leaf at the given index is correclty accepted or rejected (given a certain `threshold`).
```
Comment (Orland0x) [inline starknet/src/tests/proposal_validation_strategies/test_merkle_whitelist.cairo:24]: correctly

File: starknet/src/tests/proposal_validation_strategies/test_merkle_whitelist.cairo [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+        }
+    }
+
+    #[test]
+    #[available_gas(1000000000)]
+    fn just_enough_vp() {
+        let members = generate_n_members(20);
+        let threshold = 3_u256; // Voting power required to submit a proposal
+
+        let (contract, _) = deploy_syscall(
```
Comment (Orland0x) [inline starknet/src/tests/proposal_validation_strategies/test_merkle_whitelist.cairo:60]: could put following 10 lines into a `setup()` utility as its duplicated in every test

**Comment by pscott** [comment]
closed in favor of #496

**Pull request closed without merging**